### PR TITLE
feat: index parsing

### PR DESCRIPTION
**Pull request by JonathanStarup**
Continuing from #7982

This now works
```
def main(): Unit \ IO = region rc {
    let l = "a" :: "b" :: "c" :: Nil;
    let v = l |> List.toVector;
    let a = l |> List.toArray(rc);
    let m = l |> List.zipWithIndex |> List.toMap |> Map.toMutMap(rc);
    testIt(() -> v[1]);
    testIt(() -> v[3]);

    println("");
    testIt(() -> a[1]);
    testIt(() -> a[1] = "z");
    testIt(() -> a[1]);
    testIt(() -> a[4]);
    testIt(() -> OutOfBounds.asRes(Indexable.get(4))(a));

    println("");
    testIt(() -> m[0]);
    testIt(() -> m[0] = "q");
    testIt(() -> m[0]);
    testIt(() -> m[4]);

    println("");
    let thing = List.toArray(rc, a :: Nil);
    testIt(() -> thing[0][1]);
    testIt(() -> thing[0][1] = "pizza");
    testIt(() -> thing[0][1]);
    ()
}

def testIt(f: Unit -> a \ ef): Unit \ IO + ((ef - KeyNotFound) - OutOfBounds) with ToString[a] = {
    match KeyNotFound.asRes(OutOfBounds.asRes(f))() |> Result.flatten {
        case Ok(v) => printUnlessUnit(v)
        case Err(msg) => printUnlessUnit(msg)
    }
}
```

### CODE REVIEW

File: main/src/library/Indexable.flix [hunk 1]
```diff
@@ -41,3 +41,11 @@ pub trait Indexable[t] {
     pub def get(i: Indexable.Idx[t], t: t): Indexable.Elm[t] \ Indexable.Aef[t]
 
 }
+
+mod Indexable {
+
+    /// Alias of `get` with arguments in syntactic order.
+    pub def getSwapped(t: t, i: Indexable.Idx[t]): Indexable.Elm[t] \ Indexable.Aef[t] with Indexable[t] =
```
Comment (magnus-madsen) [inline main/src/library/Indexable.flix:48]: I assume this is to preserve evaluation order.

But what about swapping the arguments to the get/put signatures?

**Comment by magnus-madsen** [comment]
Is there any downside to flipping the arguments to get/put to be in their natural order? (I suppose only subject last?)

File: main/src/library/Indexable.flix (hunk 1 above)
Comment (JonathanStarup) [reply main/src/library/Indexable.flix:48]: we could also do that yes, just wanted to get it working. i don't think i have strong opinions on which solution

**Comment by JonathanStarup** [comment]
yeah thats the only thing.

File: main/src/ca/uwaterloo/flix/language/ast/SyntaxTree.scala [hunk 2]
```diff
@@ -182,6 +182,10 @@ object SyntaxTree {
 
       case object Apply extends Expr
 
+      case object Index extends Expr
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/SyntaxTree.scala:185]: Keep alphabetical

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 3]
```diff
@@ -815,6 +815,8 @@ object Weeder2 {
         case TreeKind.Expr.Use => visitExprUseExpr(tree)
         case TreeKind.Expr.Literal => visitLiteralExpr(tree)
         case TreeKind.Expr.Apply => visitApplyExpr(tree)
+        case TreeKind.Expr.Index => visitIndexExpr(tree)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:818]: Keep alphabetical

**Comment by magnus-madsen** [comment]
I think we should swap the arguments. We can do it in this PR or a subsequent one.

See also my two other comments.

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala (hunk 3 above)
Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:818]: I put them "alphabetical" but the rest of the list is sorted by occurrence I think?

**Comment by JonathanStarup** [comment]
swapping is a bit involved to I vote to do it in a follow up PR

**Comment by magnus-madsen** [comment]
> swapping is a bit involved to I vote to do it in a follow up PR

OK. I will merge this one then.

**Review by magnus-madsen — APPROVED** [review]

**Comment by magnus-madsen** [comment]
@JonathanStarup Can you still do the swapping -- before I let @jaschdoc loose on adding instances?

**Pull request merged**